queueing/consumer.py

```diff
@@ -62,55 +69,93 @@ def mongo_jsonify(item):
-                    item["database_name"], item["collection_name"], item["payload"]
-                )
-            else:
-                client.insert_collection_data(
-                    item["database_name"], item["collection_name"], item["payload"]
-                )
-            logging.debug("Document inserted {}.".format(item["collection_name"]))
+            item = item_queue.get(timeout=current_timeout)
+            current_timeout = QUEUE_BASE_TIMEOUT
+        except:
+            next_timeout = current_timeout * 2
+            current_timeout = (
+                QUEUE_MAX_TIMEOUT if next_timeout > QUEUE_MAX_TIMEOUT else next_timeout
+            )
+            for identifier, inserter in inserters.items():
+                inserter.flush()
+            continue
+
+        item = mongo_jsonify(item)
+        try:
+            inserters[item["identifier"]].insert(item["payload"])
         except:
             logging.warning(
-                "An exception occurs during the insertion in: {}/{}.".format(
-                    item["database_name"], item["collection_name"]
+                "An exception occurs during the insertion in {}.".format(
+                    item["identifier"]
                 )
             )
         item_queue.task_done()
 
 
-def consume_from_kafka(topic_name, worker_pool):
+def start_workers():
+    workers = []
+    for i in range(WORKER_POOL):
+        worker = threading.Thread(target=read_from_queue)
+        worker.start()
+        workers.append(worker)
+    return workers
+
+
+def heartbeat():
+    while True:
+        workers = start_workers()
+        time.sleep(HEARTBEAT_TICK)
+
+        with heartbeat_lock:
+            logging.debug("Heartbeat: A new inspection has started.")
+
+            for worker in workers:
+                worker.join()
+
+            for identifier, inserter in inserters.items():
+                if inserter.is_inactive():
+                    inserter.flush()
+                    del inserters[identifier]
+
+            logging.debug("Heartbeat: {} alive inserters.".format(len(inserters)))
+
+
+def consume_from_kafka(topic_name):
     if db_client.get_connection():
-        logging.debug("DB connection established")
+        logging.info("DB connection established.")
     else:
-        logging.error("Could not connect to the DB")
-        return
+        raise Exception("Could not connect to the DB.")
 
     consumer = connect_kafka_consumer(topic_name)
-    partitions = consumer.partitions_for_topic(topic_name)
-    workers = []
-    for i in range(worker_pool):
-        worker = threading.Thread(
-            target=read_from_queue, args=(db_client,), daemon=True
-        )
-        worker.start()
-        workers.append(worker)
+    _heartbeat = threading.Thread(target=heartbeat, daemon=True)
+    _heartbeat.start()
 
     for message in consumer:
+        if heartbeat_lock.locked():
+            heartbeat_lock.acquire()
+            heartbeat_lock.release()
```

Comment:
Why do we need to block and release? Where else is `heartbeat_lock` being locked?

Reply:
Whe have to wait for `heartbeat_lock` to be released only if it is currently locked (because a heartbeat inspection is being performed). ref https://stackoverflow.com/a/59932149